File: web/tests/reference/binary-trees/serialize-and-deserialize-binary-tree.py
class Codec:
    def serialize(self, root):
        out = []

        def walk(node):
            if node is None:
                out.append("#")
                return
            out.append(str(node.val))
            walk(node.left)
            walk(node.right)

        walk(root)
        return ",".join(out)

    def deserialize(self, data):
        tokens = iter(data.split(","))

        def build():
            token = next(tokens)
            if token == "#":
                return None
            node = TreeNode(int(token))
            node.left = build()
            node.right = build()
            return node

        return build()

File: web/tests/reference/binary-trees/serialize-and-deserialize-binary-tree.js
class Codec {
  serialize(root) {
    const out = [];
    const walk = (node) => {
      if (node === null) {
        out.push("#");
        return;
      }
      out.push(String(node.val));
      walk(node.left);
      walk(node.right);
    };
    walk(root);
    return out.join(",");
  }

  deserialize(data) {
    const tokens = data.split(",");
    let i = 0;
    const build = () => {
      const token = tokens[i++];
      if (token === "#") return null;
      const node = new TreeNode(Number(token));
      node.left = build();
      node.right = build();
      return node;
    };
    return build();
  }
}
